### PR TITLE
WhatsApp Forms: Google sheet integration

### PR DESCRIPTION
**Pull request by akanshaaa19**
target issue is #4616

### CODE REVIEW

File: lib/glific/whatsapp_forms.ex [hunk 1]
```diff
@@ -233,4 +246,118 @@ defmodule Glific.WhatsappForms do
...
+  defp extract_payload_from_child(%{
+         "on-click-action" => %{"name" => "complete", "payload" => payload}
+       }),
+       do: payload
+
+  defp extract_payload_from_child(_), do: nil
+
+  @spec extract_spreadsheet_id(String.t()) :: {:ok, String.t()} | {:error, String.t()}
+  defp extract_spreadsheet_id(url) do
+    case Regex.run(~r/\/spreadsheets\/d\/([a-zA-Z0-9-_]+)/, url) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex]: For my learning? can you please explain ,what we are doing here?

File: lib/glific/whatsapp_forms.ex [hunk 2]
```diff
@@ -233,4 +246,118 @@ defmodule Glific.WhatsappForms do
...
+  end
+
+  defp build_headers(complete_payload) do
+    default_headers = ["timestamp", "contact_id", "whatsapp_form_id"]
+    form_headers = complete_payload |> Map.keys() |> Enum.map(&to_string/1)
+
+    default_headers ++ form_headers
+  end
+
+  defp insert_headers(organization_id, spreadsheet_id, headers) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:394]: spec is missing ?

File: lib/glific/whatsapp_forms.ex [hunk 3]
```diff
@@ -233,4 +246,118 @@ defmodule Glific.WhatsappForms do
...
+
+  @spec extract_spreadsheet_id(String.t()) :: {:ok, String.t()} | {:error, String.t()}
+  defp extract_spreadsheet_id(url) do
+    case Regex.run(~r/\/spreadsheets\/d\/([a-zA-Z0-9-_]+)/, url) do
+      [_, spreadsheet_id] -> {:ok, spreadsheet_id}
+      _ -> {:error, "Invalid Google Sheets URL"}
+    end
+  end
+
+  defp build_headers(complete_payload) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:379]: spec is missing here?

File: lib/glific/whatsapp_forms.ex [hunk 4]
```diff
@@ -233,4 +246,118 @@ defmodule Glific.WhatsappForms do
...
+      {:ok, sheet} ->
+        {:ok, Map.put(attrs, :sheet_id, sheet.id)}
+
+      {:error, reason} ->
+        Logger.error("Failed to create Google Sheet for WhatsApp form: #{inspect(reason)}")
+        {:error, reason}
+    end
+  end
+
+  @spec append_headers_to_sheet(WhatsappForm.t()) :: {:ok, any()} | {:error, any()}
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:329]:
```suggestion
  @spec append_headers_to_sheet(map() | nil) :: {:ok, any()} | {:error, any()}
```

File: lib/glific/whatsapp_forms.ex [hunk 5]
```diff
@@ -233,4 +246,121 @@ defmodule Glific.WhatsappForms do
...
+      screens
+      |> Enum.find_value(&find_complete_action/1)
+
+    case payload do
+      nil -> {:error, "No complete action payload found"}
+      payload -> {:ok, payload}
+    end
+  end
+
+  defp find_complete_action(%{"layout" => %{"children" => children}}) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:360]: spec is missing here too?

File: lib/glific/whatsapp_forms.ex [hunk 6]
```diff
@@ -233,4 +246,121 @@ defmodule Glific.WhatsappForms do
...
+    children
+    |> Enum.flat_map(fn child ->
+      get_in(child, ["children"]) || []
+    end)
+    |> Enum.find_value(&extract_payload_from_child/1)
+  end
+
+  defp find_complete_action(_), do: nil
+
+  defp extract_payload_from_child(%{
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:371]: spec is missing here too?

**Comment by priyanshu6238** [comment]
Screenshot 2025-12-09 at 2 21 47 PM

The form gets accepted even when I haven’t provided edit access to the sheet. Shouldn’t we show a dialog message asking the user to provide the required write access to the sheet, or at least verify that write access has been granted?

File: lib/glific_web/schema/whatsapp_form_types.ex [hunk 7]
```diff
@@ -15,8 +17,14 @@ defmodule GlificWeb.Schema.WhatsappFormTypes do
     field :definition, :json
     field :meta_flow_id, :string
     field :categories, list_of(:string)
+    field :sheet_id, :id
     field :inserted_at, :string
     field :updated_at, :string
+
+    field :sheet, :sheet do
```
Comment (priyanshu6238) [inline lib/glific_web/schema/whatsapp_form_types.ex:26]: For my learning: can you pls explain, what is the use of dataloader(Repo) here?

File: lib/glific/whatsapp_forms.ex [hunk 8]
```diff
@@ -233,4 +246,123 @@ defmodule Glific.WhatsappForms do
...
+
+  @spec maybe_create_google_sheet(map()) ::
+          {:ok, map()} | {:error, any()}
+  defp maybe_create_google_sheet(attrs) do
+    case Map.get(attrs, :google_sheet_url) do
+      nil ->
+        {:ok, Map.put(attrs, :sheet_id, nil)}
+
+      "" ->
+        {:ok, Map.put(attrs, :sheet_id, nil)}
```
Comment (madclaws) [inline lib/glific/whatsapp_forms.ex]:
```suggestion
      url when url in ["nil", ""] ->
        {:ok, Map.put(attrs, :sheet_id, nil)}
```

File: lib/glific/whatsapp_forms.ex [hunk 9]
```diff
@@ -233,4 +246,123 @@ defmodule Glific.WhatsappForms do
...
+  defp append_headers_to_sheet(%{
+         definition: %{"screens" => screens},
+         sheet: %{url: url},
+         organization_id: organization_id
+       }) do
+    with {:ok, complete_payload} <- extract_complete_payload(screens),
+         {:ok, spreadsheet_id} <- extract_spreadsheet_id(url),
+         headers <- build_headers(complete_payload),
+         {:ok, _result} <- insert_headers(organization_id, spreadsheet_id, headers) do
+      {:ok, headers}
```
Comment (madclaws) [inline lib/glific/whatsapp_forms.ex]: Can we add test for this function, this will be useful if for some reason the form structure changes or even for better understanding. we can separate the last function which calls the ghseet service, so that we can test the header builder seperately.

**Comment by madclaws** [comment]
> Screenshot 2025-12-09 at 2 21 47 PM
> The form gets accepted even when I haven’t provided edit access to the sheet. Shouldn’t we show a dialog message asking the user to provide the required write access to the sheet, or at least verify that write access has been granted?

This should have been handled by the existing validations while creating the sheet right? @akanshaaa19

File: lib/glific/whatsapp_forms.ex [hunk 10]
```diff
@@ -233,4 +246,123 @@ defmodule Glific.WhatsappForms do
...
+  defp extract_spreadsheet_id(url) do
+    case Regex.run(~r/\/spreadsheets\/d\/([a-zA-Z0-9-_]+)/, url) do
+      [_, spreadsheet_id] -> {:ok, spreadsheet_id}
+      _ -> {:error, "Invalid Google Sheets URL"}
+    end
+  end
+
+  @spec build_headers(map()) :: list(String.t())
+  defp build_headers(complete_payload) do
+    default_headers = ["timestamp", "contact_id", "whatsapp_form_id"]
```
Comment (madclaws) [inline lib/glific/whatsapp_forms.ex]: will the contact_id useful for users instead of contact number same with wa_form_id?

Comment (akanshaaa19) [reply lib/glific/whatsapp_forms.ex]: phone number is PII so that's why i didnt add that. users can use contact_id to get the contact information is what i thought. what are your thoughts @mahajantejas @shijithkjayan @madclaws

Comment (mahajantejas) [reply lib/glific/whatsapp_forms.ex]: hey @akanshaaa19, i think we can go ahead with the phone numbers, because it is easier for the NGOs to relate, also since this is write to google sheets, the write access needs to be given only to the google service account and the sheet itself can be access controlled by the orgs

File: lib/glific/whatsapp_forms.ex (hunk 9 above)
Comment (akanshaaa19) [reply lib/glific/whatsapp_forms.ex]: i have added the test case, did not separate the functions but added the mock for it. but testing the building headers part now

File: lib/glific/whatsapp_forms.ex [hunk 11]
```diff
@@ -233,4 +252,143 @@ defmodule Glific.WhatsappForms do
...
+            {:ok, Map.put(attrs, :sheet_id, updated_sheet.id)}
+
+          {:error, reason} ->
+            Logger.error("Failed to update Google Sheet for WhatsApp form: #{inspect(reason)}")
+            {:error, reason}
+        end
+    end
+  end
+
+  @doc false
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex]: Can we add proper documentation here instead of using @doc false?

File: lib/glific/third_party/sheets/sheets.ex [hunk 12]
```diff
@@ -407,11 +433,7 @@ defmodule Glific.Sheets do
   """
   @spec execute(Action.t() | any(), FlowContext.t()) :: {FlowContext.t(), Messages.Message.t()}
   def execute(%{action_type: "WRITE"} = action, context) do
-    spreadsheet_id =
-      action.url
-      |> String.replace("https://docs.google.com/spreadsheets/d/", "")
-      |> String.split("/")
-      |> List.first()
+    spreadsheet_id = extract_spreadsheet_id(action.url)
```
Comment (priyanshu6238) [inline lib/glific/third_party/sheets/sheets.ex:459]: can you import  this function , we are getting error undefined function extract_spreadsheet_id/1 

Image

Comment (akanshaaa19) [reply lib/glific/third_party/sheets/sheets.ex:459]: its there in the same file, line 242

File: lib/glific/seeds/seeds_dev.ex [hunk 13]
```diff
@@ -2064,6 +2097,25 @@ if Code.ensure_loaded?(Faker) do
       end)
     end
 
+    @doc false
+    @spec seed_sheets(Organization.t() | nil) :: :ok
+    def seed_sheets(organization \\ nil) do
```
Comment (priyanshu6238) [inline lib/glific/seeds/seeds_dev.ex]:
```suggestion
    def seed_sheets(organization  do
```

The nil case will we handle inside the seed_whatsapp_forms we don't have to pass it here again.

File: lib/glific/seeds/seeds_dev.ex [hunk 14]
```diff
@@ -2064,6 +2097,25 @@ if Code.ensure_loaded?(Faker) do
...
+      organization = get_organization(organization)
```
Comment (priyanshu6238) [inline lib/glific/seeds/seeds_dev.ex]: We don't need to seed organization here again , it will get seeded  inside the function seed_whatsapp_forms

File: lib/glific/seeds/seeds_dev.ex (hunk 13 above)
Comment (priyanshu6238) [inline lib/glific/seeds/seeds_dev.ex]:
```suggestion
    defp seed_sheets(organization \\ nil) do
```

File: lib/glific/third_party/sheets/sheets.ex [hunk 15]
```diff
@@ -77,6 +79,29 @@ defmodule Glific.Sheets do
...
+    |> case do
+      {:ok, _} ->
+        {:ok, true}
+
+      {:error, %Tesla.Env{status: 403}} ->
+        {:error,
+         "No edit access to the Google Sheet. Please ensure the service account has editor permissions."}
+
+      {:error, reason} ->
+        {:error, "Failed to verify edit access: #{inspect(reason)}"}
```
Comment (priyanshu6238) [inline lib/glific/third_party/sheets/sheets.ex:106]: Instead of returning the entire error, let’s extract (abstract) the  reason from the response body and then pass that into the error message.

File: lib/glific/whatsapp_forms.ex [hunk 16]
```diff
@@ -233,4 +252,145 @@ defmodule Glific.WhatsappForms do
...
+          {:ok, map()} | {:error, any()}
+  defp maybe_create_google_sheet(attrs) do
+    case Map.get(attrs, :google_sheet_url) do
+      url when url in [nil, ""] ->
+        {:ok, Map.put(attrs, :sheet_id, nil)}
+
+      url when is_binary(url) ->
+        handle_google_sheet(attrs, url)
+    end
+  end
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:267]:
```suggestion
  @spec maybe_create_google_sheet(map()) ::
          {:ok, map()} | {:error, any()}
defp maybe_create_google_sheet(attrs) do
  url = Map.get(attrs, :google_sheet_url)

  if url in [nil, ""] do
    {:ok, Map.put(attrs, :sheet_id, nil)}
  else
    handle_google_sheet(attrs, url)
  end
end
```

File: lib/glific/seeds/seeds_dev.ex [hunk 17]
```diff
@@ -2064,6 +2097,23 @@ if Code.ensure_loaded?(Faker) do
       end)
     end
 
+    @doc false
```
Comment (priyanshu6238) [inline lib/glific/seeds/seeds_dev.ex]: can we remove this ,we don't need doc  for private  function

File: lib/glific/third_party/sheets/sheets.ex [hunk 18]
```diff
@@ -49,23 +49,25 @@ defmodule Glific.Sheets do
...
+      {:error, reason} ->
+        {:error, reason}
     end
   end
 
-  defp validate_sheet(attrs) do
+  defp validate_sheet(%{url: url} = _attrs) when not is_nil(url) do
     client =
       Tesla.client([
         {Tesla.Middleware.FollowRedirects, max_redirects: 5}
```
Comment (priyanshu6238) [inline lib/glific/third_party/sheets/sheets.ex:68]:
```suggestion
        {Tesla.Middleware.FollowRedirects, max_redirects: 2}
```

Why do we need max_redirects: 5 here?
Google Sheets URLs typically redirect only once or twice.

**Review by priyanshu6238 — COMMENTED** [review]
Image

the google sheet is not getting created when i publish my form? is this initially, or i am missing something here?

File: lib/glific/seeds/seeds_dev.ex [hunk 19]
```diff
@@ -2064,6 +2097,23 @@ if Code.ensure_loaded?(Faker) do
...
+          label: "Responses Sheet",
+          sheet_id: 1,
+          url: "https://docs.google.com/spreadsheets/d/#{Faker.String.base64(10)}",
+          organization_id: organization.id
+        }
+      ]
+
+      Enum.each(sheets, fn sheet ->
+        Repo.insert!(struct(Sheet, sheet))
+      end)
```
Comment (priyanshu6238) [inline lib/glific/seeds/seeds_dev.ex:2125]:
```suggestion
      end)
      :ok
```

File: lib/glific/whatsapp_forms.ex [hunk 20]
```diff
@@ -233,4 +252,145 @@ defmodule Glific.WhatsappForms do
...
+               organization_id: attrs.organization_id
+             }) do
+          {:ok, updated_sheet} ->
+            {:ok, Map.put(attrs, :sheet_id, updated_sheet.id)}
+
+          {:error, reason} ->
+            Logger.error("Failed to update Google Sheet for WhatsApp form: #{inspect(reason)}")
+            {:error, reason}
+        end
+    end
```
[the remark this answers is not in the document]
Comment (priyanshu6238) [reply lib/glific/whatsapp_forms.ex:323]: @akanshaaa19 can you fix this , it will crash at this point?

File: lib/glific/whatsapp_forms.ex [hunk 21]
```diff
@@ -233,4 +252,145 @@ defmodule Glific.WhatsappForms do
...
+      {:error, reason} ->
+        Logger.error("Failed to create Google Sheet for WhatsApp form: #{inspect(reason)}")
+        {:error, reason}
+    end
+  end
+
+  @spec update_existing_sheet(map(), String.t(), non_neg_integer()) ::
+          {:ok, map()} | {:error, any()}
+  defp update_existing_sheet(attrs, url, sheet_id) do
+    case Sheets.get_sheet!(sheet_id) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex]:
```suggestion
    case Repo.fetch_by(Sheet, %{id: sheet_id})) do
```

File: lib/glific/whatsapp_forms.ex [hunk 22]
```diff
@@ -233,4 +252,145 @@ defmodule Glific.WhatsappForms do
...
+    create_new_sheet(attrs, url)
+  end
+
+  @spec handle_google_sheet(map(), String.t()) :: {:ok, map()} | {:error, any()}
+  defp handle_google_sheet(%{operation: :update, sheet_id: sheet_id} = attrs, url)
+       when not is_nil(sheet_id) do
+    update_existing_sheet(attrs, url, sheet_id)
+  end
+
+  defp handle_google_sheet(%{operation: :update} = attrs, url) do
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex]: This looks redundant. Can we combine this with the function above line 268?
Something like:

``` 
defp handle_google_sheet(%{operation: op} = attrs, url)
     when op in [:create, :update] do
  create_new_sheet(attrs, url)
end

```

File: lib/glific/whatsapp_forms.ex [hunk 23]
```diff
@@ -233,4 +252,145 @@ defmodule Glific.WhatsappForms do
...
+    default_headers = [
+      "timestamp",
+      "contact_phone_number",
+      "whatsapp_form_id",
+      "whatsapp_form_name"
+    ]
+
+    form_headers = complete_payload |> Map.keys() |> Enum.map(&to_string/1)
+
+    default_headers ++ form_headers
```
Comment (priyanshu6238) [inline lib/glific/whatsapp_forms.ex:389]: we are adding header on top of existing header in the google sheet ,can you fix this ?

Image

File: lib/glific/third_party/sheets/sheets.ex [hunk 24]
```diff
@@ -77,6 +79,33 @@ defmodule Glific.Sheets do
     end
   end
 
+  defp validate_sheet(_attrs), do: {:ok, true}
```
Comment (priyanshu6238) [inline lib/glific/third_party/sheets/sheets.ex]: When the URL is missing and write access is not verified, we still return {:ok, true}. This seems incorrect shouldn’t we return an error in this case?

File: lib/glific/third_party/sheets/sheets.ex [hunk 25]
```diff
@@ -91,7 +120,8 @@ defmodule Glific.Sheets do
   """
   @spec update_sheet(Sheet.t(), map()) :: {:ok, Sheet.t()} | {:error, Ecto.Changeset.t()}
   def update_sheet(%Sheet{} = sheet, attrs) do
-    with {:ok, sheet} <-
+    with {:ok, true} <- validate_sheet(attrs),
```
Comment (priyanshu6238) [inline lib/glific/third_party/sheets/sheets.ex]: Are we not going to  handling the case when validate_sheet/1 fails? Shouldn’t we return the error here?

File: test/glific/third_party/whatsapp_form/whatsapp_form_test.exs [hunk 26]
```diff
@@ -48,16 +86,44 @@ defmodule Glific.ThirdParty.WhatsappForm.ApiClientTest do
...
     Tesla.Mock.mock(fn
+      %{method: :get, url: url} when is_binary(url) ->
+        %Tesla.Env{
+          status: 200,
+          body: "Key,Value\ntest1,value1\ntest2,value2"
+        }
+
+      %{method: :get, url: nil} ->
+        {:error, :invalid_url}
+
```
[the remark this answers is not in the document]
Comment (priyanshu6238) [reply test/glific/third_party/whatsapp_form/whatsapp_form_test.exs]: @akanshaaa19 can we remove this redundant mock from here?

File: test/glific/third_party/whatsapp_form/whatsapp_form_test.exs [hunk 27]
```diff
@@ -342,4 +435,75 @@ defmodule Glific.ThirdParty.WhatsappForm.ApiClientTest do
...
+              private_key: "DEFAULT PRIVATE KEY"
+            })
+        },
+        is_active: true,
+        organization_id: organization_id
+      }
+
+      Partners.create_credential(valid_attrs)
+
+      # Create a form with the complete structure from @form_json
```
Comment (priyanshu6238) [inline test/glific/third_party/whatsapp_form/whatsapp_form_test.exs:698]: can we remove all the redundant  comment down the line?

Comment (akanshaaa19) [reply test/glific/third_party/whatsapp_form/whatsapp_form_test.exs:698]: how are the comments redundant? the comments are there so that its easy to understand what the test is doing

Comment (priyanshu6238) [reply test/glific/third_party/whatsapp_form/whatsapp_form_test.exs:698]: The test cases are mostly self-explanatory, so comments may not be necessary in every case. However, adding comments is helpful when handling special scenarios or edge cases.

File: lib/glific/whatsapp_forms.ex (hunk 23 above)
Comment (akanshaaa19) [reply lib/glific/whatsapp_forms.ex:389]: can you explain this? it should add the headers only when you publish it

File: lib/glific/third_party/sheets/sheets.ex (hunk 25 above)
Comment (akanshaaa19) [reply lib/glific/third_party/sheets/sheets.ex]: we are handling, try adding a wrong link or a sheet which does not have edit access or when service account is not active- it will throw an error

**Comment by akanshaaa19** [comment]
> the google sheet is not getting created when i publish my form? is this initially, or i am missing something here? 

you need to provide a gsheet link, and on publish it will add the headers to the sheet
@priyanshu6238

File: lib/glific/whatsapp_forms.ex (hunk 23 above)
Comment (priyanshu6238) [reply lib/glific/whatsapp_forms.ex:389]: When there is some header already present in the sheet ,when publish it's rewrite it on top of the existing header

Comment (akanshaaa19) [reply lib/glific/whatsapp_forms.ex:389]: You can’t publish the form twice, so this shouldn’t happen. It’s expected to always add a new sheet, and checking for headers feels like overkill to me.

Comment (priyanshu6238) [reply lib/glific/whatsapp_forms.ex:389]: Here, we should inform the user to ensure that the Google Sheet is properly cleaned before submitting or publishing it

File: lib/glific/third_party/sheets/sheets.ex (hunk 24 above)
Comment (akanshaaa19) [reply lib/glific/third_party/sheets/sheets.ex]: validation is only required when the url is provided, and in some cases update_sheet is called without the url that's why this this added here

**Comment by priyanshu6238** [comment]
> > the google sheet is not getting created when i publish my form? is this initially, or i am missing something here?
> 
> you need to provide a gsheet link, and on publish it will add the headers to the sheet @priyanshu6238

If a user publishes the form without providing a Google Sheet URL, the form still gets published and becomes non-editable afterward. In that case, there would be no way to fix or add the sheet later.Then what we are plaining for this condition.
I thing we can avoid this issue, if we  prevent users from publishing the form until they provide a non-empty, valid Google Sheet URL.

File: lib/glific/third_party/sheets/sheets.ex (hunk 24 above)
Comment (madclaws) [inline lib/glific/third_party/sheets/sheets.ex]: This means we consider a sheet with url nil or empty as valid?, also doesn't this also means the first clause can have attrs with type WRITE, but url can be empty?

File: lib/glific/third_party/sheets/sheets.ex (hunk 18 above)
Comment (madclaws) [reply lib/glific/third_party/sheets/sheets.ex:68]: Do we have any reference on this @priyanshu6238 ?

File: lib/glific/third_party/sheets/sheets.ex [hunk 28]
```diff
@@ -49,23 +49,25 @@ defmodule Glific.Sheets do
...
   @spec validate_sheet(map()) :: {:ok, true} | {:error, String.t()}
   defp validate_sheet(%{type: "WRITE"} = attrs) do
-    GoogleSheets.fetch_credentials(attrs.organization_id)
-    |> case do
-      {:ok, _credentials} ->
-        {:ok, true}
-
-      {:error, _error} ->
+    with {:ok, _credentials} <- GoogleSheets.fetch_credentials(attrs.organization_id),
+         {:ok, true} <- check_edit_access(attrs) do
```
Comment (madclaws) [inline lib/glific/third_party/sheets/sheets.ex]: can avoid this with arm and directly call the `check_edit_access` in with body.

File: lib/glific/third_party/sheets/sheets.ex (hunk 18 above)
Comment (priyanshu6238) [reply lib/glific/third_party/sheets/sheets.ex:68]: There doesn’t seem to be any official documentation that clearly justifies this. I’m asking the same question: do we really need max_redirects: 5 here?
In Tesla, we usually rely on the default value of 5 for FollowRedirects, as mentioned in the documentation:
https://hexdocs.pm/tesla/Tesla.Middleware.FollowRedirects.html
Could you please let me know if there’s a specific reason for explicitly setting it in this case?

Comment (madclaws) [reply lib/glific/third_party/sheets/sheets.ex:68]: there might not be much reason, but not everybody is supposed to know the default of Tesla here, so this is fine even if its explicit here.

File: lib/glific/whatsapp_forms.ex (hunk 23 above)
Comment (akanshaaa19) [reply lib/glific/whatsapp_forms.ex:389]: will add as a helper text and also in the documentation

File: lib/glific/third_party/sheets/sheets.ex (hunk 24 above)
Comment (akanshaaa19) [reply lib/glific/third_party/sheets/sheets.ex]: so few places update_sheet is called without URLs because we are updating other things like the sheet sync data- that shouldn't trigger the validate function for this reason i have added a check here to validate the sheet ONLY when url is present

File: lib/glific/third_party/sheets/sheets.ex (hunk 28 above)
Comment (akanshaaa19) [reply lib/glific/third_party/sheets/sheets.ex]: done

File: lib/glific/third_party/sheets/sheets.ex [hunk 29]
```diff
@@ -91,7 +121,9 @@ defmodule Glific.Sheets do
   """
   @spec update_sheet(Sheet.t(), map()) :: {:ok, Sheet.t()} | {:error, Ecto.Changeset.t()}
   def update_sheet(%Sheet{} = sheet, attrs) do
-    with {:ok, sheet} <-
+    with {:ok, true} <-
+           if(Map.has_key?(attrs, :url), do: validate_sheet(attrs), else: {:ok, true}),
+         {:ok, sheet} <-
            sheet
            |> Sheet.changeset(attrs)
            |> Repo.update() do
```
Comment (madclaws) [inline lib/glific/third_party/sheets/sheets.ex]:
```elixir
  validated_result =
      if Map.has_key?(attrs, :url) do
        validate_sheet(attrs)
      else
        {:ok, true}
      end

    with {:ok, true} <- validated_result,
         {:ok, sheet} <-
           sheet
           |> Sheet.changeset(attrs)
           |> Repo.update() do
      # incase we update the url. Let's resync the sheet data.
      if Map.has_key?(attrs, :url), do: sync_sheet_data(sheet), else: {:ok, sheet}
    end
```

What about something like this, if inside a with clause affects readability.
